File: labs/PilhaFlex.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Celula{// define a celula
    int elemento;
    struct Celula* prox;
} Celula;

typedef struct Pilha{// define a pilha
    Celula* topo;
} Pilha;

Pilha* criarPilha(){
    Pilha* p = (Pilha*)malloc(sizeof(Pilha));
    p->topo = NULL;
    return p;
}

void empilhar(Pilha* p, int x){
    Celula* nova = (Celula*)malloc(sizeof(Celula));
    nova->elemento = x;
    nova->prox = p->topo;// aponta para o antigo topo
    p->topo = nova;// atualiza o topo
}

int desempilhar(Pilha* p){
    if(p->topo == NULL){
        printf("-1\n");// pilha vazia
        return -1;
    }
    Celula* tmp = p->topo;
    int elemento = tmp->elemento;
    p->topo = tmp->prox;// movi o topo
    free(tmp);
    printf("%d\n", elemento);
    return elemento;
}

void mostrar(Pilha* p){
    if(p->topo == NULL){
        printf("V");// vazia
    } else{
        for(Celula* i = p->topo; i != NULL; i = i->prox){
            printf("%d ", i->elemento);
        }
        printf("\n");
    }
}

int pesquisar(Pilha* p, int x){
    for(Celula* i = p->topo; i != NULL; i = i->prox){
        if(i->elemento == x)
            return 1;// encontrado
    }
    return 0;// nao encontro
}
int main(){
    Pilha* pilha = criarPilha();
    char comando;
    int num;

    while(scanf(" %c", &comando) != EOF) {
        switch(comando){
            case 'E':// empilhar
                scanf("%d", &num);
                empilhar(pilha, num);
                break;
            case 'D':
                desempilhar(pilha);
                break;
            case 'M':// pesquisa
                scanf("%d", &num);
                if(pesquisar(pilha,num))
                    printf("S\n");
                else
                    printf("N\n");
                break;
            case 'P'://mostra
                mostrar(pilha);
                break;
        }
    }
    return 0;
}

File: labs/PilhaFlex.java
import java.util.Scanner;

class Celula {
    public int elemento;
    public Celula prox;

    public Celula(int x){
        this.elemento = x;
        this.prox = null;
    }

    public Celula(){
        this(0);
    }
}

class Pilha{
    private Celula topo;

    public Pilha(){
        topo = new Celula();// celula cabeça
    }

    public void inserir(int x){
        Celula tmp = new Celula(x);
        tmp.prox = topo.prox;
        topo.prox = tmp;
    }

    public int remover(){
        if(topo.prox == null){
            System.out.println(-1);
            return -1;
        } else{
            Celula tmp = topo.prox;
            int elemento = tmp.elemento;
            topo.prox = tmp.prox;
            tmp.prox = null;
            System.out.println(elemento);
            return elemento;
        }
    }
    
    public void mostrar(){
        if(topo.prox == null){
            System.out.println("V");
        } else{
            for(Celula i = topo.prox; i != null; i = i.prox){
                System.out.print(i.elemento + " ");
            }
            System.out.println();
        }
    }

    public boolean pesquisa(int x){
        boolean existe = false;
        for(Celula i = topo.prox; i != null; i = i.prox){
            if(i.elemento == x){
                existe = true;
            }
        }
        return existe;
    }
}

public class PilhaFlex{
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        Pilha pilha = new Pilha();

        while(sc.hasNext()){
            char comando = sc.next().charAt(0);

            switch(comando){
                case 'E':// desempilhar
                    int num = sc.nextInt();
                    pilha.inserir(num);
                    break;

                case 'D': // desempilhar 
                    pilha.remover();
                    break;
                case 'M':
                    pilha.mostrar();
                    break;
                case 'P':
                    int valor = sc.nextInt();
                    if(pilha.pesquisa(valor))
                        System.out.println("S");
                    else
                        System.out.println("N");
                    break;
            }
        }
        sc.close();
    }
}